cni-plugin/linkerd-cni-validation/src/main.rs

```diff
@@ -0,0 +1,170 @@
+use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+
+use anyhow::{anyhow, Context, Result};
+use clap::Parser;
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpListener, TcpStream},
+};
+use tracing::{debug, error, info, Instrument};
+use tracing_subscriber::prelude::*;
+use tracing_subscriber::EnvFilter;
+
+#[derive(Parser)]
+#[clap(version)]
+struct Args {
+    #[clap(long, default_value = "4140")]
+    outbound_port: u16,
+    #[clap(
+        parse(try_from_str),
+        long,
+        env = "LINKERD_VALIDATION_LOG_LEVEL",
+        default_value = "debug"
+    )]
+    log_level: EnvFilter,
+    #[clap(parse(try_from_str), long, env = "KUBERNETES_SERVICE_HOST")]
+    target_host: Ipv4Addr,
+    #[clap(parse(try_from_str), long, env = "KUBERNETES_SERVICE_PORT")]
+    target_port: u16,
+}
+
+static REDIRECT_RESPONSE: &str = "REDIRECTION SUCCESSFUL";
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> anyhow::Result<()> {
+    let Args {
+        outbound_port,
+        log_level,
+        target_host,
+        target_port,
+    } = Args::parse();
+
+    tracing_subscriber::registry()
+        .with(log_level)
+        .with(tracing_subscriber::fmt::layer())
+        .try_init()?;
+
+    let target_addr = &format!("{}:{}", target_host, target_port);
+    info!(original_dst = %target_addr, outbound_redirect = %outbound_port, "validating outbound traffic is redirected to proxy outbound port");
+
+    let (shutdown_tx, shutdown_rx) = drain::channel();
+    let (ready_tx, ready_rx) = tokio::sync::oneshot::channel::<Result<()>>();
+
+    let srv_handle = tokio::spawn(async move {
+        let listen_addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), outbound_port);
+        let listener = TcpListener::bind(listen_addr)
+            .await
+            .expect(&format!("Failed to bind server to {}", listen_addr));
+        info!(server_addr=%listen_addr, "listening for incoming connections");
+
+        if let Err(_) = ready_tx.send(Ok(())) {
+            error!(server_addr=%listen_addr, "failed to send 'ready' signal, receiver dropped");
+            return Err::<(), anyhow::Error>(anyhow!(
+                "failed to bind server: listen_addr={}",
+                listen_addr
+            ));
+        }
+
+        let handler = tokio::spawn(async move {
+            loop {
+                let (mut stream, client_addr) = listener
+                    .accept()
+                    .await
+                    .expect("failed to establish connection");
+                info!(%client_addr, "accepted connection");
+                tokio::spawn(async move {
+                    let bytes = REDIRECT_RESPONSE.as_bytes();
+                    match stream.write_all(&bytes).await {
+                        Ok(_) => debug!(%client_addr, "wrote {:?} bytes", &bytes.len()),
+                        Err(e) => error!(%client_addr, "failed to write bytes to client: {:?}", e),
```

Comment:
nit: consider using structured fields in these events:
```suggestion
                        Ok(_) => debug!(%client_addr, written_bytes = bytes.len()),
                        Err(error) => error!(%client_addr, %error, "failed to write bytes to client"),
```